### PR TITLE
feat(riverpod_lint): added avoid use ref inside dispose

### PR DESCRIPTION
**Pull request by LeonardoRosaa**
It's a good practice to avoid using the `WidgetRef` in the `dispose` method to avoid some problems although it isn't every usage that creates some problems.

Identifies the dispose method by name and the `@override` annotation once a time the `WidgetRef` is only used in the `State` class.

Reference: #2542

### CODE REVIEW

**Comment by rrousselGit** [comment]
Hello!
The code isn't formatted. Could you fix that?

Also we would need tests. See the `riverpod_lint_flutter_test` folder

**Comment by LeonardoRosaa** [comment]
> Hello! The code isn't formatted. Could you fix that?
> 
> Also we would need tests. See the `riverpod_lint_flutter_test` folder

Yes. I would write tests but I got some errors after executing the `flutter test` command:

```
00:02 +0: /Users/leonardo/www/riverpod/packages/riverpod_lint_flutter_test/test/ensuire_build_test.dart: ensure_build                                                                
Running: `git rev-parse --show-toplevel` in `/Users/leonardo/www/riverpod/packages/riverpod_lint_flutter_test`
00:02 +1: /Users/leonardo/www/riverpod/packages/riverpod_lint_flutter_test/test/goldens/fixes/provider_dependencies_test.dart: Verify that @riverpod classes extend the generated typedef
[exception: PathNotFoundException(path=/Users/leonardo/flutter/bin/cache/artifacts/engine/lib/_internal/sdk_library_metadata/lib/libraries.dart; message=Cannot open file)][stackTrace: #0      _PhysicalFile.readAsStringSync (package:analyzer/file_system/physical_file_system.dart:159:7)
#1      FolderBasedDartSdk.initialLibraryMap (package:analyzer/src/dart/sdk/sdk.dart:521:41)
#2      new FolderBasedDartSdk (package:analyzer/src/dart/sdk/sdk.dart:420:18)
#3      ContextBuilderImpl._createSdk (package:analyzer/src/dart/analysis/context_builder.dart:192:21)
#4      ContextBuilderImpl.createContext (package:analyzer/src/dart/analysis/context_builder.dart:104:15)
#5      new AnalysisContextCollectionImpl (package:analyzer/src/dart/analysis/analysis_context_collection.dart:95:36)
#6      _createAnalysisContext (package:analyzer/dart/analysis/utilities.dart:127:42)
#7      resolveFile2 (package:analyzer/dart/analysis/utilities.dart:117:7)
#8      main.<anonymous closure> (file:///Users/leonardo/www/riverpod/packages/riverpod_lint_flutter_test/test/goldens/fixes/provider_dependencies_test.dart:22:28)
#9      testGolden.<anonymous closure> (file:///Users/leonardo/www/riverpod/packages/riverpod_lint_flutter_test/test/golden.dart:29:31)
#10     Declarer.test.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/declarer.dart:215:19)
<asynchronous suspension>
#11     StackZoneSpecification._registerUnaryCallback.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:124:15)
<asynchronous suspension>
]
[exception: PathNotFoundException(path=/Users/leonardo/flutter/bin/cache/artifacts/engine/lib/_internal/libraries.dart; message=Cannot open file)][stackTrace: #0      _PhysicalFile.readAsStringSync (package:analyzer/file_system/physical_file_system.dart:159:7)
#1      FolderBasedDartSdk.initialLibraryMap (package:analyzer/src/dart/sdk/sdk.dart:521:41)
#2      new FolderBasedDartSdk (package:analyzer/src/dart/sdk/sdk.dart:420:18)
#3      ContextBuilderImpl._createSdk (package:analyzer/src/dart/analysis/context_builder.dart:192:21)
#4      ContextBuilderImpl.createContext (package:analyzer/src/dart/analysis/context_builder.dart:104:15)
#5      new AnalysisContextCollectionImpl (package:analyzer/src/dart/analysis/analysis_context_collection.dart:95:36)
#6      _createAnalysisContext (package:analyzer/dart/analysis/utilities.dart:127:42)
#7      resolveFile2 (package:analyzer/dart/analysis/utilities.dart:117:7)
#8      main.<anonymous closure> (file:///Users/leonardo/www/riverpod/packages/riverpod_lint_flutter_test/test/goldens/fixes/provider_dependencies_test.dart:22:28)
#9      testGolden.<anonymous closure> (file:///Users/leonardo/www/riverpod/packages/riverpod_lint_flutter_test/test/golden.dart:29:31)
#10     Declarer.test.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/declarer.dart:215:19)
<asynchronous suspension>
#11     StackZoneSpecification._registerUnaryCallback.<anonymous closure> (package:stack_trace/src/stack_zone_specification.dart:124:15)
<asynchronous suspension>
]
```
Flutter version:
```
Flutter 3.10.5 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 796c8ef792 (3 weeks ago) • 2023-06-13 15:51:02 -0700
Engine • revision 45f6e00911
Tools • Dart 3.0.5 • DevTools 2.23.1
```

**Comment by AhmedLSayed9** [comment]
I got the same error on my [PR](https://github.com/rrousselGit/riverpod/pull/2695) and I'm stuck on this.

**Comment by rrousselGit** [comment]
Not sure how you got that error. The CI is running tests and works just fine.

**Comment by LeonardoRosaa** [comment]
> Not sure how you got that error. The CI is running tests and works just fine.

Do you run these tests only in the CI?

**Comment by LeonardoRosaa** [comment]
Hello @rrousselGit,

I am not sure that I wrote the right tests but can you check?

Thanks.

**Comment by AhmedLSayed9** [comment]
@LeonardoRosaa  Did you write the golden test manually?

**Comment by LeonardoRosaa** [comment]
> @LeonardoRosaa Did you write the golden test manually?

Yes, I did.

**Comment by AhmedLSayed9** [comment]
I see.
I think I'd go with that as I can't run tests too. Thanks!

File: packages/riverpod_lint_flutter_test/test/goldens/lints/avoid_use_ref_inside_dispose.dart [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  @override
+  ConsumerState<ConsumerStatefulWidget> createState() => _MyWidgetState();
+}
+
+class _MyWidgetState extends ConsumerState<MyWidget> {
+  @override
+  void dispose() {
+    // ignore: avoid_use_ref_inside_dispose
+    ref.read(provider);
+    // ignore: avoid_use_ref_inside_dispose
```
Comment (rrousselGit) [inline packages/riverpod_lint_flutter_test/test/goldens/lints/avoid_use_ref_inside_dispose.dart:18]: Don't use //ignore
Use //expect_lint

This will fail if the warning is missing, which ignore doesn't do

Comment (rrousselGit) [inline packages/riverpod_lint_flutter_test/test/goldens/lints/avoid_use_ref_inside_dispose.dart:18]:
```suggestion
    // expect_lint: avoid_use_ref_inside_dispose
    ref.read(provider);
    // expect_lint: avoid_use_ref_inside_dispose
```

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+const overrideAnnotation = 'override';
+const disposeMethod = 'dispose';
+
+class AvoidUseRefInsideDispose extends RiverpodLintRule {
+  const AvoidUseRefInsideDispose() : super(code: _code);
+
+  static const _code = LintCode(
+    name: 'avoid_use_ref_inside_dispose',
+    problemMessage: "Avoid using 'Ref' in the dispose method.",
```
Comment (rrousselGit) [inline packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart:17]: I think the following name would be more logical

```suggestion
    name: 'avoid_ref_inside_state_dispose',
```

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      final targetType = node.realTarget?.staticType;
+
+      if (targetType == null) return;
+
+      if (widgetRefType.isAssignableFromType(targetType)) {
+        final ancestor = node.thisOrAncestorMatching((method) {
+          if (method is MethodDeclaration) {
+            final hasOverride = method.metadata.firstWhereOrNull((element) {
+              return element.name.name == overrideAnnotation;
+            });
```
Comment (rrousselGit) [inline packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart:36]: I don't think it makes sense to check the existence of `@override`

File: packages/riverpod_lint_flutter_test/test/goldens/lints/avoid_use_ref_inside_dispose.dart [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import 'package:flutter/widgets.dart';
+import 'package:hooks_riverpod/hooks_riverpod.dart';
+
+final provider = Provider((ref) => 0);
+
+class MyWidget extends ConsumerStatefulWidget {
```
Comment (rrousselGit) [inline packages/riverpod_lint_flutter_test/test/goldens/lints/avoid_use_ref_inside_dispose.dart:6]: The following should not trigger a warning:

```suggestion
class PlainClass with ChangeNotifier {
  PlainClass(this.ref);
  final WidgetRef ref;
  @override
   void dispose() {
     ref.read(provider);
   }
}
```

**Comment by rrousselGit** [comment]
Thanks for the test! We'll probably need a few more cases to make sure we have 100% coverage.  
I've already suggested one. There may be more needed. I'd suggest looking at the core and see if anything isn't tested

**Comment by LeonardoRosaa** [comment]
> Thanks for the test! We'll probably need a few more cases to make sure we have 100% coverage. I've already suggested one. There may be more needed. I'd suggest looking at the core and see if anything isn't tested

I did some improvements. Can you check it?

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    CustomLintResolver resolver,
+    ErrorReporter reporter,
+    CustomLintContext context,
+  ) {
+    context.registry.addMethodInvocation((node) {
+      final targetType = node.realTarget?.staticType;
+
+      if (targetType == null) return;
+
+      if (widgetRefType.isAssignableFromType(targetType)) {
```
Comment (rrousselGit) [inline packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart:29]: Could you use early returns instead of nested conditions?

```suggestion
      if (!widgetRefType.isAssignableFromType(targetType)) return;
```

Comment (rrousselGit) [reply packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart:29]: Could you also use the same principle with those thisOrAncestorMatching to remove all the nesting?

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        final ancestor = node.thisOrAncestorMatching((method) {
+          if (method is MethodDeclaration &&
+              method.name.lexeme == disposeMethod) {
+            /// This [thisOrAncestorMatching] is to look for an ancestor of the
+            /// [dispose] method found
+            final classe = method.thisOrAncestorMatching((element) {
```
Comment (rrousselGit) [inline packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart:35]:
```suggestion
            final classDeclaration = method.thisOrAncestorMatching((element) {
```

**Comment by rrousselGit** [comment]
Could you document this lint in the README?

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart (hunk 5 above)
Comment (LeonardoRosaa) [reply packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart]: I did some changes following the early return principle.

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart (hunk 6 above)
Comment (LeonardoRosaa) [reply packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart]: Done!

**Comment by LeonardoRosaa** [comment]
> Could you document this lint in the README?

I wrote some documentation. But, can you explain in more detail why to avoid using ref in the dispose method?

File: packages/riverpod_lint/README.md [hunk 7]
```diff
@@ -531,6 +532,61 @@ class Example extends Anything {
...
+    ref.read(provider).doSomething();
+
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Container();
+  }
+}
```
Comment (rrousselGit) [inline packages/riverpod_lint/README.md:587]: Let's trim unrelated code

```suggestion
class _MyWidgetState extends ConsumerState<MyWidget> {
  @override
  void dispose() {
    // Do not use 'ref' in the dispose method
    ref.read(provider).doSomething();
    super.dispose();
  }

  // ...
}
```

File: packages/riverpod_lint/README.md [hunk 8]
```diff
@@ -531,6 +532,61 @@ class Example extends Anything {
...
+    return WillPopScope(
+      onWillPop: () async {
+        ref.read(provider).doSomething();
+        return true;
+      },
+      child: Container(),
+    );
+  }
+}
+```
```
Comment (rrousselGit) [inline packages/riverpod_lint/README.md:562]: This doesn't seem to add too much value. Let's remove it

````suggestion
**Good**:

```dart
class MyWidget extends ConsumerStatefulWidget {
  const MyWidget({super.key});

  @override
  ConsumerState<ConsumerStatefulWidget> createState() => _MyWidgetState();
}

class _MyWidgetState extends ConsumerState<MyWidget> {
 
  @override
  Widget build(BuildContext context) {
    return WillPopScope(
      onWillPop: () async {
        ref.read(provider).doSomething();
        return true;
      },
      child: Container(),
    );
  }
}
```
````

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  void run(
+    CustomLintResolver resolver,
+    ErrorReporter reporter,
+    CustomLintContext context,
+  ) {
+    context.registry.addMethodInvocation((node) {
+      final targetType = node.realTarget?.staticType;
+
+      if (targetType == null) return;
+
```
Comment (rrousselGit) [inline packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart:28]:
```suggestion
      if (targetType == null) return;
```

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+      if (ancestor != null) {
+        reporter.reportErrorForNode(_code, node);
+      }
+    });
+  }
+
+  /// Looking for the ConsumeState class ancestor
+  /// into the [node] parent.
+  AstNode? _findConsumeStateClass(AstNode node) {
```
Comment (rrousselGit) [inline packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart:48]:
```suggestion
  /// Looking for the ConsumerState class ancestor
  /// into the [node] parent.
  AstNode? _findConsumerStateClass(AstNode node) {
```

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+      if (ancestor != null) {
+        reporter.reportErrorForNode(_code, node);
+      }
+    });
+  }
+
+  /// Looking for the ConsumerState class ancestor
+  /// into the [node] parent.
+  AstNode? _findConsumerStateClass(AstNode node) {
+    final classeDeclaration = node.parent?.thisOrAncestorMatching((node) {
```
Comment (rrousselGit) [inline packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart:49]:
```suggestion
    return node.parent?.thisOrAncestorMatching((node) {
```

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+      /// Looking for the class which is a [ConsumeState]
+      final extendsClause = node.extendsClause;
+      if (extendsClause == null) return false;
+      final extendsType = extendsClause.superclass.type;
+      if (extendsType == null) return false;
+
+      return consumerStateType.isExactlyType(extendsType);
+    });
+
+    return classeDeclaration;
```
Comment (rrousselGit) [inline packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart:61]:
```suggestion
```

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+import '../riverpod_custom_lint.dart';
+
+const disposeMethod = 'dispose';
+
+class AvoidUseRefInsideDispose extends RiverpodLintRule {
+  const AvoidUseRefInsideDispose() : super(code: _code);
+
+  static const _code = LintCode(
+    name: 'avoid_ref_inside_state_dispose',
+    problemMessage: "Avoid using 'Ref' in the dispose method.",
```
Comment (rrousselGit) [inline packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart:15]:
```suggestion
    problemMessage: "Avoid using 'Ref' inside State.dispose.",
```

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    context.registry.addMethodInvocation((node) {
+      final targetType = node.realTarget?.staticType;
+      if (targetType == null) return;
+      if (!widgetRefType.isAssignableFromType(targetType)) return;
+
+      final ancestor = node.thisOrAncestorMatching((method) {
+        final isMethodCalledDispose =
+            method is MethodDeclaration && method.name.lexeme == disposeMethod;
+
+        if (!isMethodCalledDispose) return false;
```
Comment (rrousselGit) [inline packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart:33]:
```suggestion
        final isDisposeMethod =
            method is MethodDeclaration && method.name.lexeme == disposeMethod;

        if (!isDisposeMethod) return false;
```

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  ) {
+    context.registry.addMethodInvocation((node) {
+      final targetType = node.realTarget?.staticType;
+      if (targetType == null) return;
+      if (!widgetRefType.isAssignableFromType(targetType)) return;
+
+      final ancestor = node.thisOrAncestorMatching((method) {
+        final isDisposeMethod =
+            method is MethodDeclaration && method.name.lexeme == disposeMethod;
+
```
Comment (rrousselGit) [inline packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart:32]:
```suggestion
```

File: packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart [hunk 16]
```diff
@@ -0,0 +1,58 @@
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/error/listener.dart';
+import 'package:custom_lint_builder/custom_lint_builder.dart';
+import 'package:riverpod_analyzer_utils/riverpod_analyzer_utils.dart';
+
+import '../riverpod_custom_lint.dart';
+
+const disposeMethod = 'dispose';
+
+class AvoidUseRefInsideDispose extends RiverpodLintRule {
```
Comment (rrousselGit) [inline packages/riverpod_lint/lib/src/lints/avoid_use_ref_inside_dispose.dart:10]: Could you rename the class and file name to match the lint code? avoid_ref_inside_state_dispose

**Review by rrousselGit — APPROVED** [review]
LGTM once the lint class & filename are renamed

Thanks for this :)

**Comment by LeonardoRosaa** [comment]
> LGTM once the lint class & filename are renamed
> 
> Thanks for this :)

Done :)

**Comment by rrousselGit** [comment]
Epic. Thanks!
